core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -857,6 +858,364 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::ThreadVectorRangeBoundariesStruct<
+                          Iter> const& nextStatus,
+    Policy const& policy, Lambda const& lambda, ReducerValueType& val,
+    Args... args) {
+  lambda(args..., val);
+}
+
+template <MDTeamRangeParThread ParThread, MDTeamRangeParVector ParVector,
+          int CurrentRank, int TotalRank, int ParThreadRank, int ParVectorRank,
+          Iterate Iter, typename Policy, typename Lambda,
+          typename ReducerValueType, typename... Args>
+KOKKOS_INLINE_FUNCTION void nested_loop(
+    MDTeamRangeStatus<MDTeamRangeLastRank::NotLastRank, ParThread,
+                      ParVector> const& status,
+    MDTeamRangeNextStatus<CurrentRank, TotalRank, ParThreadRank, ParVectorRank,
+                          Iter>,
+    Policy const& policy, Lambda const& lambda, ReducerValueType& val,
+    Args... args) {
+  constexpr int nextRank = CurrentRank + (Iter == Iterate::Right ? 1 : -1);
+  using NextStatus = MDTeamRangeNextStatus<nextRank, TotalRank, ParThreadRank,
+                                           ParVectorRank, Iter>;
+
+  // FIXME Check if recursively calling this is efficient enough compared to an
+  // iterative case
+
+  // This recursively processes ranks from [0..TotalRank-1]
+  // args... is passed by value because it should always be ints
+  parallel_for(
+      nested_policy(status, policy.team, policy.boundaries[CurrentRank]),
+      [&](int const& i) {
+        if constexpr (Iter == Iterate::Right) {
+          nested_loop(typename NextStatus::type(), NextStatus(), policy, lambda,
+                      val, args..., i);
+        } else {
+          nested_loop(typename NextStatus::type(), NextStatus(), policy, lambda,
+                      val, i, args...);
+        }
+      });
+}
+
+template <MDTeamRangeThreadAndVector threadAndVector, typename MDTeamPolicy,
+          typename Lambda, typename ReductionValueType, typename Status>
+KOKKOS_INLINE_FUNCTION void md_parallel_impl(MDTeamPolicy const& policy,
+                                             Lambda const& lambda,
+                                             Status const& rangeStatus,
+                                             ReductionValueType& val) {
+  constexpr auto TotalRank = MDTeamPolicy::total_rank;
+  constexpr auto Iter      = MDTeamPolicy::iter;
+  using TeamHandle         = typename MDTeamPolicy::team_handle_type;
+
+  static_assert(MDTeamPolicy::total_rank >= 2 && MDTeamPolicy::total_rank <= 8);
+
+  using par_rank_t =
+      Impl::ParallelRank<TotalRank, Iter, typename TeamHandle::execution_space,
+                         threadAndVector>;
+
+  constexpr int startRank = (Iter == Iterate::Right) ? 0 : (TotalRank - 1);
+
+  using md_team_iter =
+      Impl::MDTeamRangeNextStatus<startRank, TotalRank, par_rank_t::par_rt,
+                                  par_rank_t::invalid, Iter>;
+
+  nested_loop(rangeStatus, md_team_iter(), policy, lambda, val);
+}
+
+}  // namespace Impl
+
+template <Iterate Iter, int TotalRank, typename TeamHandle>
```

Comment:
Maybe we should take `Kokkos::Rank<>` here instead of Iterate and TotalRank to make this have the same interface kind of like MDRangePolicy?

Reply:
One of the reasons that we moved away from `Kokkos::Rank`, even in our previous implementation, was because `Kokkos::Rank` is supporting only up to 6 ranks. I also do like the consistency that this change would have with `MDRangePolicy`. So I will change this to use `Kokkos::Rank`, if 6 ranks is enough.

https://github.com/kokkos/kokkos/blob/d19aab9981a2c447e832a7b4eb7b16992328fb14/core/src/Kokkos_Rank.hpp#L54-L60

Reply:
Changed to use `Kokkos::Rank` in [c534937](https://github.com/kokkos/kokkos/pull/5238/commits/c5349378887bbbf53a29cd76eb21095cdd18b83d)

Will make a separate pr to handle `Kokkos::Rank` requiring ranks to be less than 7.